packages/react-router/src/searchMiddleware.ts

```diff
@@ -9,13 +9,14 @@ export function retainSearchParams<TSearchSchema extends object>(
   return ({ search, next }) => {
     const result = next(search)
     if (keys === true) {
-      return { ...search, ...result }
+      return { ...result, ...search }
```

Comment:
I don't get this. `retainSearchParams` should only carry over search params that are NOT part of e.g. a `<Link search={...}>`. but this change here would mean that if the search params already contained `foo: 'hello'`, then the value from `<Link search={{foo:'bar'}}>` would be ignored

Reply:
Shouldn't it? The value is retained, whatever it's initialized to. Imagine you go to route x with param=p1, then goto x/y with search param=p2, this would means the param is p2 when ur on x/y and reset back to p1 when go on x. This means the value is not retained on x/* route right? If this is not the intention, I might have misunderstood the point.

Reply:
no, the intention of this search middleware is to keep search params around so they continue to live on with their previous value unless a Link overrides them

Reply:
Ooh, ok then. Feel free to close this PR, cuz I'm not home for next 2 or 3 days.

* next(search) returns the validated result (including what's in the links search prop) so it's not gonna be as easy as this fix.

Reply:
no rush, just convert it to draft for the time being